### PR TITLE
Faceting

### PR DESCRIPTION
**Pull request by emeeks**
Implement dimension and metric faceting based off of sending an array of `facets` as a `metadata.dx` prop.

* UI for metric faceting that automatically creates facets based on available metrics
* Capability to send more sophisticated facets by sending custom payloads (no UI currently, but you could pass facets that filter based on dim or instantiate different chart types or more complicated settings)
* Cleaned up logic where the new capabilities revealed issues (such as shuffling the color values of dimensions or moving the grid generation into `updateChart` and out of `render()`)
* Moved the legend out of individual charts for facets and spawned at the end, along with a shared color setting across charts
* Suppressed `VizControls` when faceting was enabled
* introduced some functionality for suppressing controls though exposed no UI to activate yet
* Added documentation

### CODE REVIEW

File: src/Documentation.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+The `data` prop expects a well-formatted data object:
+
+```ts
+type Data = {
+      schema: { fields: [{ name: string, type: string }...], primaryKey: Array<?string> },
+      data: [{ key: value },...]
+    }
+```
+
+Below is an example of a data explorer with an example of a valid data object.
```
Comment (captainsafia) [inline src/Documentation.md]:
```suggestion
Below is an example of a data explorer with a valid data object.
```

File: src/Documentation.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+import DataExplorer, { Viz, Toolbar } from "@nteract/data-explorer";
+
+<DataExplorer data={largeVizData}>
+  <Toolbar />
+  <Viz />
+  <Toolbar />
+</DataExplorer>;
+```
+
+### No Toolbar
```
Comment (captainsafia) [inline src/Documentation.md]: Didn't know this. Nifty!

File: src/charts/bar.tsx [hunk 3]
```diff
@@ -20,7 +20,9 @@ interface BarOptions {
 export const semioticBarChart = (
   data: Dx.Datapoint[],
   schema: Dx.Schema,
-  options: BarOptions
+  options: BarOptions,
+  colorHashOverride?: Object,
```
Comment (captainsafia) [inline src/charts/bar.tsx]: Is there a more specific interface we can use for this instead of `Object`?

Comment (emeeks) [reply src/charts/bar.tsx]: Updated to `{ key?: string }`

**Pull request merged**